drivers/media/platform/msm/sde/rotator/sde_rotator_r3.c

```diff
@@ -2760,15 +2760,23 @@ static struct sde_rot_hw_resource *sde_hw_rotator_alloc_ext(
 		resinfo->hw.max_active = SDE_HW_ROT_REGDMA_TOTAL_CTX - 1;
 
 		if (!test_bit(SDE_CAPS_HW_TIMESTAMP, mdata->sde_caps_map) &&
-				resinfo->rot->swts_buf.mapped == false)
-			sde_hw_rotator_swts_create(resinfo->rot);
-
-		if (resinfo->rot->swts_buf.mapped == false) {
-			ret = sde_hw_rotator_swts_map(resinfo->rot);
-			if (ret) {
-				SDEROT_ERR("swts buffer map failed\n");
-				devm_kfree(&mgr->pdev->dev, resinfo);
-				return NULL;
+				resinfo->rot->swts_buf.mapped == false) {
+
+			/*
+			 * IOMMU MDP CTX split between ROT and MDP needs
+			 * special handling: never free the endire context
+			 * or MDP gets freed as well (and havoc happens)
+			 */
+			if (test_bit(SDE_CAPS_SPLIT_CTX,
+					mdata->sde_caps_map)) {
+				ret = sde_hw_rotator_swts_map(resinfo->rot);
+				if (ret) {
+					SDEROT_ERR("swts buffer map failed\n");
+					devm_kfree(&mgr->pdev->dev, resinfo);
+					return NULL;
+				}
+			} else {
+				sde_hw_rotator_swts_create(resinfo->rot);
```

Comment:
If I understand this correctly devices with `SDE_CAPS_SPLIT_CTX` never enter `sde_hw_rotator_swts_create`, hence leaving `srcp_dma_buf` unassigned, passing a `NULL` to `sde_smmu_map_dma_buf` inside `sde_hw_rotator_swts_map`: that seems totally wrong. Same applies to `srcp_table` and `srcp_attachment`.

Have a look at the way I addressed it here: https://github.com/sonyxperiadev/kernel/pull/2197/commits/327d8ab63d291cd9d63d4bceaf302d4503d9b282

A proper fix might include checking for previously created/mapped resources in `_create`, as discussed [here](https://github.com/sonyxperiadev/kernel/pull/2197#issuecomment-587752635).

Reply:
It's fine :)))